homeassistant/components/legrand_rflc/__init__.py

```diff
@@ -0,0 +1,81 @@
+"""The Legrand RFLC integration.
+
+https://www.legrand.us/solutions/smart-lighting/radio-frequency-lighting-controls
+"""
+
+import asyncio
+from collections.abc import Mapping
+from typing import Final
+
+import lc7001.aio
+
+from homeassistant import config_entries
+from homeassistant.config_entries import SOURCE_REAUTH
+from homeassistant.const import CONF_AUTHENTICATION, CONF_HOST, CONF_PORT
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+
+PLATFORMS: Final = ["light"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict) -> bool:
+    """Set up Legrand RFLC integration."""
+    hass.data[DOMAIN] = {}
+    return True
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: config_entries.ConfigEntry
+) -> bool:
+    """Set up Legrand LC7001 from a config entry."""
+    entry_id = entry.entry_id
+    data = entry.data
+    host = data[CONF_HOST]
+    kwargs = {}
+    if CONF_AUTHENTICATION in data:
+        kwargs["key"] = bytes.fromhex(data[CONF_AUTHENTICATION])
+    if CONF_PORT in data:  # for testing only (server emulation on localhost)
+        kwargs["port"] = data[CONF_PORT]
+    hass.data[DOMAIN][entry_id] = hub = lc7001.aio.Hub(host, **kwargs)
+
+    async def setup_platforms() -> None:
+        hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    async def _reauth() -> None:
+        await hass.config_entries.async_unload(entry_id)
+        await hass.config_entries.flow.async_init(
```

Comment:
My async_setup_entry creates a task to asynchronously interact with the lc7001 hub.
From this task, the nested reauth() callback will be awaited whenever a UNAUTHENTICATED event occurs.
My reauth logic needs to unload this entry which will end up canceling the very task that reauth() is running in.
I can't do that.
So, I return immediately after starting a nested _reauth task to do the work.
The new task running _reauth() could await async_unload and then call the async_start_reauth helper function.
The downside of this is that it creates yet another task and my async_step_reauth function would have to lookup the entry from the entry_id in the context.

I think it would be keeping it as is.

Reply:
https://github.com/home-assistant/core/pull/53430/commits/285d70c1c873e5a1ee52666ffc7eab7c4ead23f2